policy-controller/k8s/status/src/index.rs

```diff
@@ -0,0 +1,222 @@
+use crate::{
+    http_route::{ParentReference, RouteBinding},
+    resource_id::ResourceId,
+};
+use ahash::{AHashMap as HashMap, AHashSet as HashSet};
+use chrono::offset::Utc;
+use linkerd_policy_controller_k8s_api::{self as k8s, gateway, ResourceExt};
+use parking_lot::RwLock;
+use std::{collections::hash_map::Entry, sync::Arc};
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+
+const POLICY_API_GROUP: &str = "policy.linkerd.io";
+const POLICY_API_VERSION: &str = "policy.linkerd.io/v1alpha1";
+const STATUS_CONTROLLER_NAME: &str = "policy.linkerd.io/status-controller";
+
+pub type SharedIndex = Arc<RwLock<Index>>;
+
+pub struct Controller {
+    client: k8s::Client,
+    updates: UnboundedReceiver<Update>,
+}
+
+pub struct Index {
+    updates: UnboundedSender<Update>,
+
+    http_routes: HashMap<ResourceId, RouteBinding>,
+    servers: HashSet<ResourceId>,
+}
+
+pub struct Update {
+    id: ResourceId,
+    patch: k8s::Patch<serde_json::Value>,
+}
+
+impl Controller {
+    pub fn new(client: k8s::Client, updates: UnboundedReceiver<Update>) -> Self {
+        Self { client, updates }
+    }
+
+    pub async fn process_updates(mut self) {
+        let patch_params = k8s::PatchParams::apply("policy.linkerd.io");
+
+        while let Some(Update { id, patch }) = self.updates.recv().await {
+            let api =
+                k8s::Api::<k8s::policy::HttpRoute>::namespaced(self.client.clone(), &id.namespace);
+            if let Err(error) = api.patch_status(&id.name, &patch_params, &patch).await {
+                tracing::error!(namespace = %id.namespace, name = %id.name, %error, "Failed to patch HTTPRoute");
+            }
+        }
+    }
+}
+
+impl Index {
+    pub fn shared(updates: UnboundedSender<Update>) -> SharedIndex {
+        Arc::new(RwLock::new(Self {
+            updates,
+            http_routes: HashMap::new(),
+            servers: HashSet::new(),
+        }))
+    }
+
+    fn update_http_route(&mut self, id: ResourceId, binding: RouteBinding) -> bool {
+        match self.http_routes.entry(id) {
+            Entry::Vacant(entry) => {
+                entry.insert(binding);
+            }
+            Entry::Occupied(mut entry) => {
+                if *entry.get() == binding {
+                    return false;
+                }
+                entry.insert(binding);
+            }
+        }
+        true
+    }
+
+    fn make_http_route_patch(
+        &self,
+        id: &ResourceId,
+        binding: &RouteBinding,
+    ) -> k8s::Patch<serde_json::Value> {
+        let accepting_servers: Vec<ResourceId> = self
+            .servers
+            .iter()
+            .filter(|server| binding.selects_server(server))
+            .cloned()
+            .collect();
+        let parent_statuses = binding
```

Comment:
maybe worth adding a TODO comment about skipping creating a patch for httproute resources which already have the desired status?

Reply:
I'm cautious to add a comment about doing this. It's worth confirming as a follow-up how this works, but I think with the `k8s::Patch::Merge` type we're okay patching with something that may already be there.

In order to determine what statuses are already there, we either have to change the index to also store the status as part of the `RouteBinding`, or we have to get the HTTPRoute with an API call which moves all this back to asynchronous work.

Both of those options feel like they'd add a lot to this change, and not save that much on patching with the merge type that we already use.